File: tools/render-all-formats.py
#!/usr/bin/env python3

import sys
import os
import subprocess
import yaml

format_keep = {
    'pdf': 'tex',
    'typst': 'typ'
}
format_output = {
    'pdf': 'pdf',
    'html': 'html',
    'typst': 'pdf',
    'dashboard': 'html',
    'docx': 'docx',
    'pptx': 'pptx'
}

def extract_metadata_from_file(file):
    with open(file, 'r') as f:
        lines = f.readlines()
    start = None
    end = None
    for i, line in enumerate(lines):
        if line.strip() == '---':
            if start is None:
                start = i
            else:
                end = i
                metadata = yaml.load(''.join(lines[start+1:end]), Loader=yaml.SafeLoader)
                return metadata
    raise ValueError('No metadata found in file %s' % file)


if len(sys.argv) < 3:
    print('usage: render-all-formats.py [--dryrun] output-root doc.qmd ...')
    print('  creates output-root/doc/format/...')
    print('  output-root should be empty')

dryrun = False
argc = 1
if sys.argv[argc] == '--dryrun':
    dryrun = True
    argc += 1
elif sys.argv[argc].startswith('--'):
    print('unsupported option', sys.argv[argc])
    sys.exit(1)

output_root = sys.argv[argc]
qmdfiles = sys.argv[argc+1:]

for qmdfile in qmdfiles:
    if not qmdfile.endswith('.qmd'):
        print('expecting only .qmd files, skipping', qmdfile)
        continue

    print(qmdfile)
    qmdbase = qmdfile[:-4]
    qmdroot = qmdbase.split('/')[-1]
    meta = extract_metadata_from_file(qmdfile)
    for format, spec in meta['format'].items():
        if not (outext := format_output.get(format)):
            print(f'unsupported format {format}, skipping')
            continue
        outdir = '/'.join([output_root, qmdroot, format])
        os.makedirs(outdir, exist_ok=True)
        metadata = []
        if keepext := format_keep.get(format):
            metadata = [
                '-M', f'keep-{keepext}:true',
                '-M', f'output-ext:pdf']
        qcmd = [
            'quarto',
            'render',
            qmdfile,
            '-t', format,
            '-M', 'keep-md:true',
            *metadata
        ]
        print(' '.join(qcmd))
        if not dryrun:
            result = subprocess.call(qcmd)
            if result != 0:
                sys.exit(1)
        movefiles = [
            f'{qmdbase}.{outext}',
            f'{qmdbase}.{format}.md'
        ]
        if keepext:
            movefiles.append(f'{qmdbase}.{keepext}')
        movefiles.append(f'{qmdbase}_files')
        for movefile in movefiles:
            filename = movefile.split('/')[-1]
            dest = f'{outdir}/{filename}'
            print(f'mv {movefile} {dest}')
            if not dryrun:
                try:
                    os.rename(movefile, dest)
                except FileNotFoundError:
                    print('... not found')

File: tools/render-all-formats.ts
#!/usr/bin/env -S deno run --unstable

import * as fs from 'stdlib/fs';
import * as yaml from 'stdlib/yaml';
import * as path from 'stdlib/path';

const formatKeep: Record<string, string> = {
  'pdf': 'tex',
  'typst': 'typ',
};

const formatOutput: Record<string, string> = {
  'pdf': 'pdf',
  'html': 'html',
  'typst': 'pdf',
  'dashboard': 'html',
  'docx': 'docx',
  'pptx': 'pptx',
  'docusaurus-md': 'mdx',
  'revealjs': 'html',
};

async function extractMetadataFromFile(file: string): Promise<any> {
  const fileContents = await Deno.readTextFile(file);
  const lines = fileContents.split('\n');
  let start: number | null = null;
  let end: number | null = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line === '---') {
      if (start === null) {
        start = i;
      } else {
        end = i;
        const metadata = yaml.parse(lines.slice(start + 1, end).join('\n'));
        return metadata;
      }
    }
  }
  throw new Error(`No metadata found in file ${file}`);
}

async function renderAndMoveArtifacts(dryRun: boolean, outputRoot: string, qmdFile : string) {
  if (!qmdFile.endsWith('.qmd')) {
    console.log('expecting only .qmd files, skipping', qmdFile);
    return 0;
  }

  console.log(qmdFile);
  const qmdbase = path.basename(qmdFile).slice(0, -4);
  let qmddir = path.dirname(qmdFile);
  if(qmddir.includes('..') || qmdFile.startsWith('/')) {
    console.warn("Warning: currently unable to replicate absolute or .. paths. Try running from the common root directory of the files if you get AlreadyExists errors.")
    qmddir = '.'
  }
  const meta = await extractMetadataFromFile(qmdFile);

  const mdformat = meta['format'];
  if (!mdformat) {
    console.log(`does not contain format, skipping`, qmdFile);
    return 0;
  }
  const formats = typeof mdformat === 'string' ? [mdformat] : Object.keys(mdformat);

  let nprocessed = 0;
  for (const format of formats) {
    const outext = formatOutput[format];
    if (!outext) {
      console.log(`unsupported format ${format}, skipping`, qmdFile);
      continue;
    }
    const outdir = path.join(outputRoot, qmddir, qmdbase, format);
    console.log(`mkdir -p ${outdir}`);
    if (!dryRun && !await fs.exists(outdir)) {
      await Deno.mkdir(outdir, { recursive: true });
    }
    const metadata: string[] = [];
    const keepext = formatKeep[format];
    if (keepext) {
      metadata.push('-M', `keep-${keepext}:true`);
    }
    const qcmd = [
      'render',
      qmdFile,
      '-t',
      format,
      '-M',
      'keep-md:true',
      ...metadata
    ];
    console.log('quarto', ...qcmd);
    if (!dryRun) {
      const cmd = new Deno.Command('quarto', {
          args: qcmd
      });
      const output = await cmd.output();
      if (!output.success) {
          console.log(new TextDecoder().decode(output.stderr));
          Deno.exit(1);
      }
    }
    const movefiles = [
      `${qmdbase}.${outext}`,
      `${qmdbase}.${format}.md`
    ];
    if (keepext) {
      movefiles.push(`${qmdbase}.${keepext}`);
    }
    movefiles.push(`${qmdbase}_files`);
    for (const movefile of movefiles) {
      const src = path.join(qmddir, movefile);
      const dest = path.join(outdir, movefile);
      if(dryRun) {
        console.log(`mv ${src} ${dest}`);
      } else {
        try {
          await fs.move(src, dest);
          console.log(`√ mv ${src} ${dest}`);
        } catch (error) {
          if(error instanceof Deno.errors.NotFound) {
            console.log(`x mv ${src} ${dest}`);
          }
          else {
              console.error(error);
              Deno.exit(1);
          }
        }
      }
    }
    nprocessed++;
  }
  return nprocessed;
}

if (import.meta.main) {
  const startTime = performance.now();
  const args = Deno.args;
  if (args.includes('--help') || args.includes('-h') || args.filter(arg => !arg.startsWith('-')).length < 2) {
    console.log('usage: render-all-formats.ts [--dryrun] output-root doc.qmd ...');
    console.log('  creates output-root/doc/format/...');
    console.log('  output-root should be empty');
    Deno.exit(1);
  }

  let dryRun = false;
  let argc = 0;
  if (args[argc] === '--dryrun') {
    dryRun = true;
    argc += 1;
  } else if (args[argc].startsWith('--')) {
    console.log('unsupported option', args[argc]);
    Deno.exit(1);
  }

  const outputRoot = args[argc];
  const qmdFiles = args.slice(argc + 1);

  const promises = qmdFiles.map(renderAndMoveArtifacts.bind(null, dryRun, outputRoot));
  const counts = await Promise.all(promises);
  const count = counts.reduce((a, b) => a+b, 0)
  const endTime = performance.now();
  const elapsed = endTime - startTime;
  console.log(`Rendered ${count} documents in ${(elapsed/1000.).toFixed(2)} seconds`);
}
